{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b70d404c",
   "metadata": {},
   "source": [
    "# Group 1 Throw-In Project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "eb8a7d58",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting tqdm\n",
      "  Downloading tqdm-4.67.1-py3-none-any.whl.metadata (57 kB)\n",
      "Downloading tqdm-4.67.1-py3-none-any.whl (78 kB)\n",
      "Installing collected packages: tqdm\n",
      "\u001b[33m  WARNING: The script tqdm is installed in '/usr/local/python/3.12.1/bin' which is not on PATH.\n",
      "  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.\u001b[0m\u001b[33m\n",
      "\u001b[0mSuccessfully installed tqdm-4.67.1\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m25.1.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.2\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpython -m pip install --upgrade pip\u001b[0m\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Collecting statsbombpy\n",
      "  Downloading statsbombpy-1.16.0-py3-none-any.whl.metadata (63 kB)\n",
      "Requirement already satisfied: pandas in /home/codespace/.local/lib/python3.12/site-packages (from statsbombpy) (2.3.1)\n",
      "Requirement already satisfied: requests in /home/codespace/.local/lib/python3.12/site-packages (from statsbombpy) (2.32.4)\n",
      "Collecting requests-cache (from statsbombpy)\n",
      "  Downloading requests_cache-1.2.1-py3-none-any.whl.metadata (9.9 kB)\n",
      "Collecting inflect (from statsbombpy)\n",
      "  Downloading inflect-7.5.0-py3-none-any.whl.metadata (24 kB)\n",
      "Requirement already satisfied: joblib in /home/codespace/.local/lib/python3.12/site-packages (from statsbombpy) (1.5.1)\n",
      "Collecting more_itertools>=8.5.0 (from inflect->statsbombpy)\n",
      "  Downloading more_itertools-10.8.0-py3-none-any.whl.metadata (39 kB)\n",
      "Collecting typeguard>=4.0.1 (from inflect->statsbombpy)\n",
      "  Downloading typeguard-4.4.4-py3-none-any.whl.metadata (3.3 kB)\n",
      "Requirement already satisfied: typing_extensions>=4.14.0 in /home/codespace/.local/lib/python3.12/site-packages (from typeguard>=4.0.1->inflect->statsbombpy) (4.14.1)\n",
      "Requirement already satisfied: numpy>=1.26.0 in /home/codespace/.local/lib/python3.12/site-packages (from pandas->statsbombpy) (2.3.1)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /home/codespace/.local/lib/python3.12/site-packages (from pandas->statsbombpy) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/codespace/.local/lib/python3.12/site-packages (from pandas->statsbombpy) (2025.2)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /home/codespace/.local/lib/python3.12/site-packages (from pandas->statsbombpy) (2025.2)\n",
      "Requirement already satisfied: six>=1.5 in /home/codespace/.local/lib/python3.12/site-packages (from python-dateutil>=2.8.2->pandas->statsbombpy) (1.17.0)\n",
      "Requirement already satisfied: charset_normalizer<4,>=2 in /home/codespace/.local/lib/python3.12/site-packages (from requests->statsbombpy) (3.4.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/codespace/.local/lib/python3.12/site-packages (from requests->statsbombpy) (3.10)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /home/codespace/.local/lib/python3.12/site-packages (from requests->statsbombpy) (2.5.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/codespace/.local/lib/python3.12/site-packages (from requests->statsbombpy) (2025.7.9)\n",
      "Requirement already satisfied: attrs>=21.2 in /home/codespace/.local/lib/python3.12/site-packages (from requests-cache->statsbombpy) (25.3.0)\n",
      "Collecting cattrs>=22.2 (from requests-cache->statsbombpy)\n",
      "  Downloading cattrs-25.2.0-py3-none-any.whl.metadata (8.4 kB)\n",
      "Requirement already satisfied: platformdirs>=2.5 in /home/codespace/.local/lib/python3.12/site-packages (from requests-cache->statsbombpy) (4.3.8)\n",
      "Collecting url-normalize>=1.4 (from requests-cache->statsbombpy)\n",
      "  Downloading url_normalize-2.2.1-py3-none-any.whl.metadata (5.6 kB)\n",
      "Downloading statsbombpy-1.16.0-py3-none-any.whl (16 kB)\n",
      "Downloading inflect-7.5.0-py3-none-any.whl (35 kB)\n",
      "Downloading more_itertools-10.8.0-py3-none-any.whl (69 kB)\n",
      "Downloading typeguard-4.4.4-py3-none-any.whl (34 kB)\n",
      "Downloading requests_cache-1.2.1-py3-none-any.whl (61 kB)\n",
      "Downloading cattrs-25.2.0-py3-none-any.whl (70 kB)\n",
      "Downloading url_normalize-2.2.1-py3-none-any.whl (14 kB)\n",
      "Installing collected packages: url-normalize, typeguard, more_itertools, cattrs, requests-cache, inflect, statsbombpy\n",
      "\u001b[?25l\u001b[33m  WARNING: The script url-normalize is installed in '/usr/local/python/3.12.1/bin' which is not on PATH.\n",
      "  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.\u001b[0m\u001b[33m\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m7/7\u001b[0m [statsbombpy]\u001b[0m [cattrs]\n",
      "\u001b[1A\u001b[2KSuccessfully installed cattrs-25.2.0 inflect-7.5.0 more_itertools-10.8.0 requests-cache-1.2.1 statsbombpy-1.16.0 typeguard-4.4.4 url-normalize-2.2.1\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m25.1.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.2\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpython -m pip install --upgrade pip\u001b[0m\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    " %pip install tqdm\n",
    " %pip install statsbombpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "57e0ce99",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from statsbombpy import sb\n",
    "\n",
    "from typing import Optional, Iterable\n",
    "from statsbombpy import sb\n",
    "from tqdm.auto import tqdm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "972fcebc",
   "metadata": {},
   "source": [
    "### Import Womans Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9807cb4e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# 2020-21 Womens Super League (Prem for Females)\n",
    "\n",
    "WomenSL = sb.matches(competition_id=37, season_id=90) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "73073089",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>match_id</th>\n",
       "      <th>match_date</th>\n",
       "      <th>competition</th>\n",
       "      <th>season</th>\n",
       "      <th>home_team</th>\n",
       "      <th>away_team</th>\n",
       "      <th>competition_stage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3775648</td>\n",
       "      <td>2021-02-28</td>\n",
       "      <td>England - FA Women's Super League</td>\n",
       "      <td>2020/2021</td>\n",
       "      <td>Aston Villa</td>\n",
       "      <td>Arsenal WFC</td>\n",
       "      <td>Regular Season</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3775609</td>\n",
       "      <td>2021-04-28</td>\n",
       "      <td>England - FA Women's Super League</td>\n",
       "      <td>2020/2021</td>\n",
       "      <td>Arsenal WFC</td>\n",
       "      <td>West Ham United LFC</td>\n",
       "      <td>Regular Season</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3775633</td>\n",
       "      <td>2021-02-06</td>\n",
       "      <td>England - FA Women's Super League</td>\n",
       "      <td>2020/2021</td>\n",
       "      <td>Aston Villa</td>\n",
       "      <td>Tottenham Hotspur Women</td>\n",
       "      <td>Regular Season</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3775570</td>\n",
       "      <td>2021-03-28</td>\n",
       "      <td>England - FA Women's Super League</td>\n",
       "      <td>2020/2021</td>\n",
       "      <td>Brighton &amp; Hove Albion WFC</td>\n",
       "      <td>Everton LFC</td>\n",
       "      <td>Regular Season</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3775581</td>\n",
       "      <td>2021-03-28</td>\n",
       "      <td>England - FA Women's Super League</td>\n",
       "      <td>2020/2021</td>\n",
       "      <td>Chelsea FCW</td>\n",
       "      <td>Aston Villa</td>\n",
       "      <td>Regular Season</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   match_id  match_date  ...                away_team competition_stage\n",
       "0   3775648  2021-02-28  ...              Arsenal WFC    Regular Season\n",
       "1   3775609  2021-04-28  ...      West Ham United LFC    Regular Season\n",
       "2   3775633  2021-02-06  ...  Tottenham Hotspur Women    Regular Season\n",
       "3   3775570  2021-03-28  ...              Everton LFC    Regular Season\n",
       "4   3775581  2021-03-28  ...              Aston Villa    Regular Season\n",
       "\n",
       "[5 rows x 7 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Clean WomenSL\n",
    "WomenSL_Clean = WomenSL[['match_id', 'match_date', 'competition', 'season', 'home_team', 'away_team', 'competition_stage']]\n",
    "WomenSL_Clean.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8f538f4e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fetching events:   0%|          | 0/131 [00:00<?, ?it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:   1%|          | 1/131 [00:00<00:17,  7.51it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:   2%|▏         | 2/131 [00:00<00:15,  8.27it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:   2%|▏         | 3/131 [00:00<00:14,  8.66it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:   3%|▎         | 4/131 [00:00<00:14,  8.68it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:   4%|▍         | 5/131 [00:00<00:14,  8.93it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:   5%|▍         | 6/131 [00:00<00:13,  9.16it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:   5%|▌         | 7/131 [00:00<00:13,  9.33it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:   6%|▌         | 8/131 [00:00<00:13,  9.36it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:   7%|▋         | 9/131 [00:00<00:13,  9.25it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:   8%|▊         | 10/131 [00:01<00:13,  9.22it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:   8%|▊         | 11/131 [00:01<00:12,  9.24it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  10%|▉         | 13/131 [00:01<00:12,  9.64it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  11%|█         | 14/131 [00:01<00:12,  9.38it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  11%|█▏        | 15/131 [00:01<00:12,  9.52it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  12%|█▏        | 16/131 [00:01<00:12,  9.48it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  13%|█▎        | 17/131 [00:01<00:12,  9.31it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  14%|█▎        | 18/131 [00:01<00:11,  9.47it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  15%|█▍        | 19/131 [00:02<00:12,  9.32it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  15%|█▌        | 20/131 [00:02<00:11,  9.50it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  16%|█▌        | 21/131 [00:02<00:11,  9.54it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  17%|█▋        | 22/131 [00:02<00:11,  9.63it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  18%|█▊        | 24/131 [00:02<00:10,  9.83it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  19%|█▉        | 25/131 [00:02<00:10,  9.79it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  20%|█▉        | 26/131 [00:02<00:10,  9.62it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  21%|██        | 27/131 [00:02<00:11,  9.45it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  21%|██▏       | 28/131 [00:03<00:11,  9.16it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  22%|██▏       | 29/131 [00:03<00:11,  9.20it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  23%|██▎       | 30/131 [00:03<00:10,  9.35it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  24%|██▎       | 31/131 [00:03<00:10,  9.44it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  24%|██▍       | 32/131 [00:03<00:10,  9.48it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  25%|██▌       | 33/131 [00:03<00:10,  9.40it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  26%|██▌       | 34/131 [00:03<00:10,  9.35it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  27%|██▋       | 35/131 [00:03<00:10,  9.26it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  27%|██▋       | 36/131 [00:03<00:10,  9.29it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  29%|██▉       | 38/131 [00:04<00:09,  9.69it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  30%|██▉       | 39/131 [00:04<00:09,  9.48it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  31%|███       | 40/131 [00:04<00:09,  9.22it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  31%|███▏      | 41/131 [00:04<00:09,  9.11it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  32%|███▏      | 42/131 [00:04<00:09,  9.24it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  33%|███▎      | 43/131 [00:04<00:09,  9.12it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  34%|███▎      | 44/131 [00:06<00:46,  1.87it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  34%|███▍      | 45/131 [00:06<00:35,  2.44it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  35%|███▌      | 46/131 [00:06<00:27,  3.10it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  36%|███▌      | 47/131 [00:06<00:21,  3.87it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  37%|███▋      | 48/131 [00:06<00:17,  4.68it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  37%|███▋      | 49/131 [00:06<00:14,  5.54it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  38%|███▊      | 50/131 [00:06<00:12,  6.30it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  39%|███▉      | 51/131 [00:06<00:11,  6.89it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  40%|███▉      | 52/131 [00:07<00:10,  7.30it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  40%|████      | 53/131 [00:07<00:10,  7.74it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  41%|████      | 54/131 [00:07<00:09,  7.97it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  42%|████▏     | 55/131 [00:07<00:09,  8.12it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  44%|████▎     | 57/131 [00:07<00:08,  8.77it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  44%|████▍     | 58/131 [00:07<00:08,  8.87it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  45%|████▌     | 59/131 [00:07<00:07,  9.06it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  46%|████▌     | 60/131 [00:07<00:07,  9.09it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  47%|████▋     | 61/131 [00:08<00:07,  8.84it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  47%|████▋     | 62/131 [00:08<00:07,  8.88it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  48%|████▊     | 63/131 [00:08<00:07,  9.11it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  49%|████▉     | 64/131 [00:08<00:07,  8.94it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  50%|████▉     | 65/131 [00:08<00:07,  9.16it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  50%|█████     | 66/131 [00:08<00:07,  9.09it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  51%|█████     | 67/131 [00:08<00:07,  9.13it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  52%|█████▏    | 68/131 [00:08<00:06,  9.32it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  53%|█████▎    | 69/131 [00:08<00:06,  9.29it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  53%|█████▎    | 70/131 [00:09<00:06,  9.39it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  54%|█████▍    | 71/131 [00:09<00:06,  9.52it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  55%|█████▍    | 72/131 [00:09<00:06,  9.45it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  56%|█████▌    | 73/131 [00:09<00:06,  9.41it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  56%|█████▋    | 74/131 [00:09<00:06,  9.23it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  57%|█████▋    | 75/131 [00:09<00:06,  8.85it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  58%|█████▊    | 76/131 [00:09<00:06,  8.93it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  59%|█████▉    | 77/131 [00:09<00:06,  8.95it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  60%|█████▉    | 78/131 [00:09<00:05,  9.10it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  60%|██████    | 79/131 [00:10<00:05,  9.08it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  61%|██████    | 80/131 [00:10<00:05,  9.30it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  63%|██████▎   | 82/131 [00:10<00:05,  9.48it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  64%|██████▍   | 84/131 [00:10<00:04,  9.60it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  65%|██████▍   | 85/131 [00:10<00:04,  9.40it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  66%|██████▌   | 86/131 [00:10<00:04,  9.45it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  66%|██████▋   | 87/131 [00:10<00:04,  9.47it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  67%|██████▋   | 88/131 [00:10<00:04,  9.00it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  68%|██████▊   | 89/131 [00:11<00:04,  8.91it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  69%|██████▉   | 91/131 [00:11<00:04,  9.24it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  70%|███████   | 92/131 [00:11<00:04,  9.02it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  71%|███████   | 93/131 [00:11<00:04,  8.78it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  72%|███████▏  | 94/131 [00:11<00:04,  8.81it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  73%|███████▎  | 95/131 [00:11<00:04,  8.81it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  73%|███████▎  | 96/131 [00:12<00:12,  2.79it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  74%|███████▍  | 97/131 [00:12<00:09,  3.47it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  75%|███████▍  | 98/131 [00:12<00:07,  4.28it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  76%|███████▌  | 99/131 [00:13<00:06,  5.14it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  76%|███████▋  | 100/131 [00:13<00:05,  5.89it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  77%|███████▋  | 101/131 [00:13<00:04,  6.26it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  78%|███████▊  | 102/131 [00:13<00:04,  6.91it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  79%|███████▊  | 103/131 [00:13<00:03,  7.43it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  79%|███████▉  | 104/131 [00:13<00:03,  7.97it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  80%|████████  | 105/131 [00:13<00:03,  8.44it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  81%|████████  | 106/131 [00:13<00:02,  8.39it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  82%|████████▏ | 107/131 [00:13<00:02,  8.78it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  82%|████████▏ | 108/131 [00:14<00:02,  8.83it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  83%|████████▎ | 109/131 [00:14<00:02,  8.69it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  84%|████████▍ | 110/131 [00:14<00:02,  8.94it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  85%|████████▍ | 111/131 [00:14<00:02,  9.08it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  85%|████████▌ | 112/131 [00:14<00:02,  9.26it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  87%|████████▋ | 114/131 [00:14<00:01,  9.36it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  88%|████████▊ | 115/131 [00:14<00:01,  9.32it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  89%|████████▊ | 116/131 [00:14<00:01,  9.12it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  89%|████████▉ | 117/131 [00:15<00:01,  9.24it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  90%|█████████ | 118/131 [00:15<00:01,  9.30it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  91%|█████████ | 119/131 [00:15<00:01,  9.09it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  92%|█████████▏| 120/131 [00:15<00:01,  9.14it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  92%|█████████▏| 121/131 [00:15<00:01,  8.99it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  93%|█████████▎| 122/131 [00:15<00:00,  9.06it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  94%|█████████▍| 123/131 [00:15<00:00,  9.29it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  95%|█████████▍| 124/131 [00:15<00:00,  9.35it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  95%|█████████▌| 125/131 [00:15<00:00,  9.34it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  96%|█████████▌| 126/131 [00:16<00:00,  9.34it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  97%|█████████▋| 127/131 [00:16<00:00,  8.94it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  98%|█████████▊| 129/131 [00:16<00:00,  9.63it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events:  99%|█████████▉| 130/131 [00:16<00:00,  9.39it/s]C:\\Users\\Earni\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\statsbombpy\\api_client.py:21: NoAuthWarning: credentials were not supplied. open data access only\n",
      "  warnings.warn(\n",
      "Fetching events: 100%|██████████| 131/131 [00:16<00:00,  7.91it/s]\n"
     ]
    }
   ],
   "source": [
    "# Function for Matches Left Join Events\n",
    "## Looking though all the events and only keeping the ones that occured in matches present in WomenSL\n",
    "def build_events_dataset(\n",
    "    matches_df: pd.DataFrame,\n",
    "    match_id_col: str = \"match_id\",\n",
    "    save_csv: Optional[str] = None,\n",
    "    keep_columns: Optional[Iterable[str]] = None,\n",
    "    add_meta: bool = True,\n",
    "    ) -> pd.DataFrame:\n",
    "\n",
    "# Fetch events for all match_ids in `matches_df`, concatenate into one DataFrame,\n",
    "# and (optionally) merge key match metadata columns.\n",
    "\n",
    "# Parameters\n",
    "# ----------\n",
    "# matches_df : DataFrame containing at least `match_id`\n",
    "# match_id_col : name of the match id column\n",
    "# save_parquet : path to save a Parquet file (recommended)\n",
    "# save_csv : path to save a CSV file\n",
    "# keep_columns : iterable of event columns to keep (None = keep all)\n",
    "# add_meta : merge match metadata columns (season, teams, date, week, competition)\n",
    "\n",
    "# Returns\n",
    "# -------\n",
    "# DataFrame of all events with `match_id` and optional metadata.\n",
    "\n",
    "    if match_id_col not in matches_df.columns:\n",
    "        raise KeyError(f\"`{match_id_col}` not in matches_df columns\")\n",
    "\n",
    "\n",
    "    meta_cols = [\n",
    "    match_id_col,\n",
    "    \"season\", \"season_id\", \"season_name\",\n",
    "    \"competition\", \"competition_id\", \"competition_name\",\n",
    "    \"match_date\", \"match_week\",\n",
    "    \"home_team\", \"away_team\"\n",
    "    ]\n",
    "    meta = matches_df[[c for c in meta_cols if c in matches_df.columns]].drop_duplicates()\n",
    "\n",
    "    all_frames = []\n",
    "    failures = []\n",
    "\n",
    "    mids = pd.unique(matches_df[match_id_col].astype(\"int64\"))\n",
    "    for mid in tqdm(mids, desc=\"Fetching events\"):\n",
    "        try:\n",
    "            ev = sb.events(match_id=int(mid))\n",
    "            ev[match_id_col] = int(mid)\n",
    "            all_frames.append(ev)\n",
    "        except Exception as e:\n",
    "            failures.append((int(mid), str(e)))\n",
    "\n",
    "    if not all_frames:\n",
    "        raise RuntimeError(\"No events were loaded. Check your match_ids or environment.\")\n",
    "\n",
    "    events = pd.concat(all_frames, ignore_index=True)\n",
    "\n",
    "\n",
    "    if keep_columns is not None:\n",
    "        keep = list(set(keep_columns) | {match_id_col})\n",
    "        events = events[[c for c in keep if c in events.columns]]\n",
    "\n",
    "\n",
    "    if add_meta and not meta.empty:\n",
    "        events = events.merge(meta, on=match_id_col, how=\"left\")\n",
    "\n",
    "\n",
    "    if save_csv:\n",
    "        events.to_csv(save_csv, index=False)\n",
    "        print(f\"Saved events: {len(events):,} rows → {save_csv}\")\n",
    "\n",
    "    if failures:\n",
    "        print(f\"\\n Skipped {len(failures)} matches due to errors (showing first 5):\")\n",
    "    for mid, msg in failures[:5]:\n",
    "        print(f\" match_id={mid}: {msg}\")\n",
    "\n",
    "    return events\n",
    "\n",
    "\n",
    "\n",
    "events_df = build_events_dataset(\n",
    "WomenSL_Clean,\n",
    "keep_columns=None,\n",
    "add_meta=True\n",
    ")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "42fe21f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(443304, 122)\n",
      "       match_id     season                   home_team                away_team\n",
      "0       3775648  2020/2021                 Aston Villa              Arsenal WFC\n",
      "3811    3775609  2020/2021                 Arsenal WFC      West Ham United LFC\n",
      "7392    3775633  2020/2021                 Aston Villa  Tottenham Hotspur Women\n",
      "10742   3775570  2020/2021  Brighton & Hove Albion WFC              Everton LFC\n",
      "14444   3775581  2020/2021                 Chelsea FCW              Aston Villa\n"
     ]
    }
   ],
   "source": [
    "print(events_df.shape)\n",
    "print(events_df[[\"match_id\", \"season\", \"home_team\", \"away_team\"]].drop_duplicates().head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "9a9b9d34",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean events column list\n",
    "event_columns = ['id', \n",
    "                'index', \n",
    "                'location', \n",
    "                'match_id',\n",
    "                'season',\n",
    "                'competition',\n",
    "                'match_date',\n",
    "                'home_team',\n",
    "                'away_team',\n",
    "                'team', \n",
    "                'interception_outcome', \n",
    "                'minute', \n",
    "                'second',\n",
    "                'pass_angle', \n",
    "                'pass_body_part', \n",
    "                'pass_cross', \n",
    "                'pass_deflected', \n",
    "                'pass_end_location', \n",
    "                'pass_goal_assist', \n",
    "                'pass_height', \n",
    "                'pass_length',\n",
    "                'pass_outcome',\n",
    "                'pass_recipient',\n",
    "                'pass_recipient_id',\n",
    "                'pass_type',\n",
    "                'play_pattern',\n",
    "                'player',\n",
    "                'player_id', \n",
    "                'position',\n",
    "                'possession_team',\n",
    "                'possession_team_id',\n",
    "                'related_events',\n",
    "                'type',\n",
    "                'under_pressure'\n",
    "                ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "0332075a",
   "metadata": {},
   "outputs": [],
   "source": [
    "events_clean = events_df[event_columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "ce459a1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# select rows where play_pattern is exactly 'From Throw In' and keep original events_df index\n",
    "throws = events_clean[events_clean['play_pattern'] == 'From Throw In'].copy()\n",
    "throws['event_idx'] = throws.index  # preserve original events_df index for lookup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "37131542",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Provide a robust helper to convert an event's minute/second to absolute seconds\n",
    "def _event_seconds(ev_row):\n",
    "    \"\"\"Return seconds from start of match for an event row (safe to missing values).\n",
    "    ev_row is typically a pandas Series representing an events_df row.\n",
    "    \"\"\"\n",
    "    # support dict-like or Series-like access\n",
    "    if hasattr(ev_row, 'get'):\n",
    "        minute = ev_row.get('minute', 0)\n",
    "        second = ev_row.get('second', 0)\n",
    "    else:\n",
    "        minute = ev_row['minute'] if 'minute' in ev_row else 0\n",
    "        second = ev_row['second'] if 'second' in ev_row else 0\n",
    "    try:\n",
    "        m = int(minute) if minute is not None else 0\n",
    "    except Exception:\n",
    "        m = 0\n",
    "    try:\n",
    "        s = int(second) if second is not None else 0\n",
    "    except Exception:\n",
    "        s = 0\n",
    "    return m * 60 + s\n",
    "\n",
    "# quick test (optional): uncomment to run a small check when executing this cell\n",
    "# sample = {'minute': 12, 'second': 34}\n",
    "# print(_event_seconds(sample))  # -> 754"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e90bd816",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build next_actions up to 3 actions (if not already present) and compute throw-level possession metrics\n",
    "rows = []\n",
    "for _, throw in throws.iterrows():\n",
    "    mid = throw['match_id']\n",
    "    throw_idx = throw['event_idx']\n",
    "    throw_team = throw.get('team') if isinstance(throw.get('team'), dict) else throw.get('team')\n",
    "    throw_name = throw_team.get('name') if isinstance(throw_team, dict) else throw_team\n",
    "    throw_secs = throw.get('throw_seconds', None)\n",
    "    later = events_df[(events_df['match_id'] == mid) & (events_df.index > throw_idx)].sort_index().head(3)\n",
    "    for action_number, (_, act) in enumerate(later.iterrows(), start=1):\n",
    "        action_team_obj = act.get('team')\n",
    "        action_team = action_team_obj.get('name') if isinstance(action_team_obj, dict) else action_team_obj\n",
    "        is_pass = isinstance(act.get('pass'), dict)\n",
    "        pass_success = None\n",
    "        if is_pass:\n",
    "            p = act.get('pass')\n",
    "            pass_success = ('outcome' not in p) if isinstance(p, dict) else None\n",
    "        is_shot = isinstance(act.get('shot'), dict)\n",
    "        is_goal = False\n",
    "        if is_shot:\n",
    "            s = act.get('shot')\n",
    "            out = s.get('outcome') if isinstance(s, dict) else None\n",
    "            if isinstance(out, dict):\n",
    "                is_goal = (out.get('name') == 'Goal')\n",
    "        action_secs = _event_seconds(act)\n",
    "        time_from_throw = pd.NA\n",
    "        try:\n",
    "            if (throw_secs is not None) and (action_secs is not None):\n",
    "                time_from_throw = action_secs - throw_secs\n",
    "        except Exception:\n",
    "            time_from_throw = pd.NA\n",
    "        same_team = (action_team == throw_name)\n",
    "        rows.append({\n",
    "            'throw_event_idx': throw_idx,\n",
    "            'match_id': mid,\n",
    "            'throw_team': throw_name,\n",
    "            'action_number': action_number,\n",
    "            'action_index': act.name,\n",
    "            'action_type': (act.get('type').get('name') if isinstance(act.get('type'), dict) else act.get('type')),\n",
    "            'action_team': action_team,\n",
    "            'action_seconds': action_secs,\n",
    "            'time_from_throw': time_from_throw,\n",
    "            'same_team_as_throw': same_team,\n",
    "            'is_pass': bool(is_pass),\n",
    "            'pass_success': pass_success,\n",
    "            'is_shot': bool(is_shot),\n",
    "            'is_goal': bool(is_goal)\n",
    "        })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf4732f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "next_actions = pd.DataFrame(rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0af2c5c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Aggregate per throw_event_idx (i.e. per throw-in) to determine possession retention and other signals\n",
    "def agg_throw(group):\n",
    "    # group is rows for one throw-in, actions ordered by action_number\n",
    "    group = group.sort_values('action_number')\n",
    "    n_actions = len(group)\n",
    "    # possession retained flags for first 1/2/3 actions (only True if those actions exist and are same team)\n",
    "    retained1 = bool(group[group['action_number'] == 1]['same_team_as_throw'].all()) if n_actions >= 1 else pd.NA\n",
    "    retained2 = bool(group[group['action_number'] <= 2]['same_team_as_throw'].all()) if n_actions >= 2 else pd.NA\n",
    "    retained3 = bool(group[group['action_number'] <= 3]['same_team_as_throw'].all()) if n_actions >= 3 else pd.NA\n",
    "    # where possession first lost (action_number) or NA if never lost in observed actions\n",
    "    lost_actions = group[group['same_team_as_throw'] == False]['action_number']\n",
    "    possession_lost_at = int(lost_actions.iloc[0]) if not lost_actions.empty else pd.NA\n",
    "    # any shot/goal by throw team within observed actions\n",
    "    any_shot = any((group['is_shot']) & (group['action_team'] == group['throw_team'].iloc[0]))\n",
    "    any_goal = any((group['is_goal']) & (group['action_team'] == group['throw_team'].iloc[0]))\n",
    "    # time to first shot by throw team\n",
    "    shot_times = group[(group['is_shot']) & (group['action_team'] == group['throw_team'].iloc[0])]['time_from_throw'].dropna()\n",
    "    time_to_first_shot = float(shot_times.min()) if not shot_times.empty else pd.NA\n",
    "    mean_time = float(group['time_from_throw'].dropna().mean()) if not group['time_from_throw'].dropna().empty else pd.NA\n",
    "    return pd.Series({\n",
    "        'n_actions_observed': n_actions,\n",
    "        'possession_retained_1': retained1,\n",
    "        'possession_retained_2': retained2,\n",
    "        'possession_retained_3': retained3,\n",
    "        'possession_lost_at': possession_lost_at,\n",
    "        'any_shot_by_throw_team': any_shot,\n",
    "        'any_goal_by_throw_team': any_goal,\n",
    "        'time_to_first_shot': time_to_first_shot,\n",
    "        'mean_time_from_throw': mean_time\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4b40e46",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<positron-console-cell-38>:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n"
     ]
    }
   ],
   "source": [
    "throw_sequences = next_actions.groupby('throw_event_idx').apply(agg_throw).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffa4f201",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"['team'] not in index\"",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)\n",
      "Cell \u001b[1;32mIn[39], line 4\u001b[0m\n",
      "\u001b[0;32m      2\u001b[0m meta \u001b[38;5;241m=\u001b[39m throws[[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mevent_idx\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmatch_id\u001b[39m\u001b[38;5;124m'\u001b[39m]]\u001b[38;5;241m.\u001b[39mrename(columns\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mevent_idx\u001b[39m\u001b[38;5;124m'\u001b[39m:\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mthrow_event_idx\u001b[39m\u001b[38;5;124m'\u001b[39m})\n",
      "\u001b[0;32m      3\u001b[0m throw_sequences \u001b[38;5;241m=\u001b[39m throw_sequences\u001b[38;5;241m.\u001b[39mmerge(meta, on\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mthrow_event_idx\u001b[39m\u001b[38;5;124m'\u001b[39m, how\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mleft\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[1;32m----> 4\u001b[0m throw_sequences \u001b[38;5;241m=\u001b[39m throw_sequences\u001b[38;5;241m.\u001b[39mmerge(\u001b[43mthrows\u001b[49m\u001b[43m[\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mevent_idx\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mteam\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mrename(columns\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mevent_idx\u001b[39m\u001b[38;5;124m'\u001b[39m:\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mthrow_event_idx\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mteam\u001b[39m\u001b[38;5;124m'\u001b[39m:\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mthrow_team_obj\u001b[39m\u001b[38;5;124m'\u001b[39m}), on\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mthrow_event_idx\u001b[39m\u001b[38;5;124m'\u001b[39m, how\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mleft\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[0;32m      5\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21m_team_name\u001b[39m(obj):\n",
      "\u001b[0;32m      6\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\pandas\\core\\frame.py:4108\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[1;34m(self, key)\u001b[0m\n",
      "\u001b[0;32m   4106\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m is_iterator(key):\n",
      "\u001b[0;32m   4107\u001b[0m         key \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(key)\n",
      "\u001b[1;32m-> 4108\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_indexer_strict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcolumns\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m[\u001b[38;5;241m1\u001b[39m]\n",
      "\u001b[0;32m   4110\u001b[0m \u001b[38;5;66;03m# take() does not accept boolean indexers\u001b[39;00m\n",
      "\u001b[0;32m   4111\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(indexer, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdtype\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m) \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mbool\u001b[39m:\n",
      "\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\pandas\\core\\indexes\\base.py:6200\u001b[0m, in \u001b[0;36mIndex._get_indexer_strict\u001b[1;34m(self, key, axis_name)\u001b[0m\n",
      "\u001b[0;32m   6197\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "\u001b[0;32m   6198\u001b[0m     keyarr, indexer, new_indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reindex_non_unique(keyarr)\n",
      "\u001b[1;32m-> 6200\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_raise_if_missing\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkeyarr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis_name\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;32m   6202\u001b[0m keyarr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtake(indexer)\n",
      "\u001b[0;32m   6203\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(key, Index):\n",
      "\u001b[0;32m   6204\u001b[0m     \u001b[38;5;66;03m# GH 42790 - Preserve name from an Index\u001b[39;00m\n",
      "\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python313\\site-packages\\pandas\\core\\indexes\\base.py:6252\u001b[0m, in \u001b[0;36mIndex._raise_if_missing\u001b[1;34m(self, key, indexer, axis_name)\u001b[0m\n",
      "\u001b[0;32m   6249\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNone of [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mkey\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m] are in the [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00maxis_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m]\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;32m   6251\u001b[0m not_found \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(ensure_index(key)[missing_mask\u001b[38;5;241m.\u001b[39mnonzero()[\u001b[38;5;241m0\u001b[39m]]\u001b[38;5;241m.\u001b[39munique())\n",
      "\u001b[1;32m-> 6252\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mnot_found\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m not in index\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\n",
      "\u001b[1;31mKeyError\u001b[0m: \"['team'] not in index\""
     ]
    }
   ],
   "source": [
    "# join throw metadata (match_id, throw_team) from one action row if available or from throws df\n",
    "meta = throws[['event_idx','match_id']].rename(columns={'event_idx':'throw_event_idx'})\n",
    "throw_sequences = throw_sequences.merge(meta, on='throw_event_idx', how='left')\n",
    "throw_sequences = throw_sequences.merge(throws[['event_idx','team']].rename(columns={'event_idx':'throw_event_idx','team':'throw_team_obj'}), on='throw_event_idx', how='left')\n",
    "def _team_name(obj):\n",
    "    try:\n",
    "        return obj.get('name') if isinstance(obj, dict) else obj\n",
    "    except Exception:\n",
    "        return obj\n",
    "throw_sequences['throw_team'] = throw_sequences['throw_team_obj'].apply(_team_name)\n",
    "throw_sequences = throw_sequences.drop(columns=['throw_team_obj'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca8076c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example evaluator function: tune thresholds and rules here\n",
    "def evaluate_throw_in(seq_row, require_retained_actions: int = 3, allow_shot_success: bool = True):\n",
    "    # success if possession retained for required actions OR a shot/goal by throw team occurs in the window\n",
    "    retained_key = f'possession_retained_{require_retained_actions}'\n",
    "    retained = seq_row.get(retained_key)\n",
    "    if retained is True:\n",
    "        return True\n",
    "    if allow_shot_success and seq_row.get('any_shot_by_throw_team'):\n",
    "        return True\n",
    "    return False\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1dfbf4a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "throw_sequences['success_default'] = throw_sequences.apply(lambda r: evaluate_throw_in(r, require_retained_actions=3, allow_shot_success=True), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "faf1e4e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show samples\n",
    "print('Action-level (next_actions) sample:')\n",
    "display(next_actions.head(20))\n",
    "print('\\nThrow-level sequences (possession & shot metrics):')\n",
    "display(throw_sequences.head(20))\n",
    "\n",
    "# You can now filter throw_sequences by success_default or tune evaluate_throw_in to your criteria."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
